File: backend/app/schemas/restitution.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

RestitutionStatus = Literal["pending", "in_progress", "completed", "cancelled"]


class RestitutionRead(BaseModel):
    id: uuid.UUID
    match_id: uuid.UUID
    meeting_location: str | None
    meeting_latitude: float | None
    meeting_longitude: float | None
    proof_photos: list[str]
    status: str
    handoff_confirmed_by_owner: bool = False
    handoff_confirmed_by_finder: bool = False
    rating_by_owner: int | None
    rating_by_finder: int | None
    comment_by_owner: str | None = None
    comment_by_finder: str | None = None
    completed_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class RestitutionUpdate(BaseModel):
    meeting_location: str | None = None
    meeting_latitude: float | None = None
    meeting_longitude: float | None = None
    status: RestitutionStatus | None = None


class RestitutionRating(BaseModel):
    """Submit a post-restitution rating (1-5 stars) + optional comment."""
    rating: int = Field(..., ge=1, le=5)
    comment: str | None = Field(None, max_length=140)

File: mobile/lib/features/restitution/models/restitution.dart
class Restitution {
  final String id;
  final String matchId;
  final String? meetingLocation;
  final List<String> proofPhotos;
  final String status; // pending | in_progress | completed | cancelled
  final bool handoffConfirmedByOwner;
  final bool handoffConfirmedByFinder;
  final int? ratingByOwner;
  final int? ratingByFinder;
  final DateTime? completedAt;
  final DateTime? createdAt;

  const Restitution({
    required this.id,
    required this.matchId,
    this.meetingLocation,
    this.proofPhotos = const [],
    required this.status,
    this.handoffConfirmedByOwner = false,
    this.handoffConfirmedByFinder = false,
    this.ratingByOwner,
    this.ratingByFinder,
    this.completedAt,
    this.createdAt,
  });

  bool get isCompleted => status == 'completed';
  bool get bothConfirmed =>
      handoffConfirmedByOwner && handoffConfirmedByFinder;

  factory Restitution.fromJson(Map<String, dynamic> j) => Restitution(
        id: j['id'] as String,
        matchId: j['match_id'] as String,
        meetingLocation: j['meeting_location'] as String?,
        proofPhotos:
            (j['proof_photos'] as List?)?.cast<String>() ?? const [],
        status: j['status'] as String? ?? 'pending',
        handoffConfirmedByOwner: (j['handoff_confirmed_by_owner'] ??
            j['confirmed_by_owner']) as bool? ??
            false,
        handoffConfirmedByFinder: (j['handoff_confirmed_by_finder'] ??
            j['confirmed_by_finder']) as bool? ??
            false,
        ratingByOwner: j['rating_by_owner'] as int?,
        ratingByFinder: j['rating_by_finder'] as int?,
        completedAt: j['completed_at'] != null
            ? DateTime.tryParse(j['completed_at'] as String)
            : null,
        createdAt: j['created_at'] != null
            ? DateTime.tryParse(j['created_at'] as String)
            : null,
      );
}
